src/main/java/ch/njol/skript/lang/FunctionParser.java

```diff
@@ -0,0 +1,407 @@
+import org.skriptlang.skript.lang.function.FunctionReference.Argument;
+import org.skriptlang.skript.lang.function.FunctionReference.ArgumentType;
+import org.skriptlang.skript.lang.function.Parameter;
+
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * A class containing the methods to parse an expression to a {@link FunctionReference}.
+ *
+ * @param context The context of parsing.
+ * @param flags   The active parsing flags.
+ */
+record FunctionParser(ParseContext context, int flags) {
+
+	private final static Pattern FUNCTION_CALL_PATTERN =
+		Pattern.compile("(?<name>[\\p{IsAlphabetic}_][\\p{IsAlphabetic}\\d_]*)\\((?<args>.*)\\)");
+
+	/**
+	 * Attempts to parse {@code expr} as a function reference.
+	 *
+	 * @param <T> The return type of the function.
+	 * @return A {@link FunctionReference} if a function is found, or {@code null} if none is found.
+	 */
+	<T> FunctionReference<T> parseFunctionReference(String expr) {
+		if (context != ParseContext.DEFAULT && context != ParseContext.EVENT) {
+			return null;
+		}
+
+		try (ParseLogHandler log = SkriptLogger.startParseLogHandler()) {
+			if (!expr.endsWith(")")) {
+				log.printLog();
+				return null;
+			}
+
+			Matcher matcher = FUNCTION_CALL_PATTERN.matcher(expr);
+			if (!matcher.matches()) {
+				log.printLog();
+				return null;
+			}
+
+			String functionName = matcher.group("name");
+			String args = matcher.group("args");
+
+			// Check for incorrect quotes, e.g. "myFunction() + otherFunction()" being parsed as one function
+			// See https://github.com/SkriptLang/Skript/issues/1532
+			for (int i = 0; i < args.length(); i = SkriptParser.next(args, i, context)) {
+				if (i == -1) {
+					log.printLog();
+					return null;
+				}
+			}
+
+			if ((flags & SkriptParser.PARSE_EXPRESSIONS) == 0) {
+				Skript.error("Functions cannot be used here (or there is a problem with your arguments).");
+				log.printError();
+				return null;
+			}
+
+			return parseFunctionReference(functionName, args, log);
+		}
+	}
+
+	/**
+	 * Attempts to parse a function reference.
+	 *
+	 * @param name The function name.
+	 * @param args The passed arguments as a string to the function.
+	 * @param <T>  The return type of the function.
+	 * @return A {@link FunctionReference} if a function is found, or {@code null} if none is found.
+	 */
+	private <T> FunctionReference<T> parseFunctionReference(String name, String args, ParseLogHandler log) {
+		FunctionReference.Argument<String>[] arguments = new FunctionArgumentParser(args).getArguments();
+
+		String namespace;
+		if (ParserInstance.get().isActive()) {
+			namespace = ParserInstance.get().getCurrentScript().getConfig().getFileName();
+		} else {
+			namespace = null;
+		}
+
+		// try to find a matching signature to get which types to parse args with
+		Set<Signature<?>> options = FunctionRegistry.getRegistry().getSignatures(namespace, name);
```

Comment:
can't we do this prior to arg parsing to save some work?

Reply:
im not sure i understand what work would be saved by moving this prior to arg parsing. args are parsed as strings. their actual return type is only found when trying to match a signature in the `options` set.

Reply:
All the work done in FunctionArgumentParser#parse()